rasa/shared/core/domain.py

```diff
@@ -1875,23 +1890,51 @@ def _validate_slot_mappings(forms: Union[Dict, List]) -> None:
             raise InvalidDomain(
                 f"If you use the deprecated list syntax for forms, "
                 f"all form names have to be strings. Please see "
-                f"{rasa.shared.constants.DOCS_URL_FORMS} for more information."
+                f"{DOCS_URL_FORMS} for more information."
             )
 
         return
 
     if not isinstance(forms, dict):
         raise InvalidDomain("Forms have to be specified as dictionary.")
 
-    for form_name, slots in forms.items():
-        if slots is None:
+    for form_name, form_data in forms.items():
+        if form_data is None:
             continue
 
+        if not isinstance(form_data, Dict):
+            raise InvalidDomain(
+                f"The contents of form '{form_name}' were specified "
+                f"as '{type(form_data)}'. They need to be specified "
+                f"as dictionary. Please see {DOCS_URL_FORMS} "
+                f"for more information."
+            )
+
+        if REQUIRED_SLOTS_KEY in form_data:
+            slots = forms[form_name].get(REQUIRED_SLOTS_KEY)
+        else:
+            rasa.shared.utils.io.raise_deprecation_warning(
+                f"The definition of slot mappings in your form"
+                f" should always be preceded by the keyword {REQUIRED_SLOTS_KEY}."
+                f" The lack of this keyword will be deprecated in"
+                f" Rasa Open Source 3.0.0. Please see "
+                f"{DOCS_URL_FORMS} for more information.",
+            )
+            slots = form_data
+
+        if IGNORED_INTENTS in form_data and REQUIRED_SLOTS_KEY not in form_data:
+            raise InvalidDomain(
+                f"If you use the {IGNORED_INTENTS} parameter in your form, then"
+                f" the keyword {REQUIRED_SLOTS_KEY} should precede the definition"
+                f" of your slot mappings. Please see"
+                f" {DOCS_URL_FORMS} for more information."
+            )
```

Comment:
I think this check should go before the one in line 1913 because if we actually raise an exception then there is no reason to also bother the user with a warning.

Reply:
That's a screenshot which shows both warnings.
<img width="1127" alt="Screenshot 2021-04-22 at 14 04 18" src="https://user-images.githubusercontent.com/7667273/115711335-ac2b0880-a373-11eb-9081-9da2f6be4bfb.png">

Reply:
True!